src/lib/tls/tls_session_manager.cpp

```diff
@@ -0,0 +1,208 @@
+            //    The ticket_lifetime_hint field contains a hint from the
+            //    server about how long the ticket should be stored. [...]
+            //    A client SHOULD delete the ticket and associated state when
+            //    the time expires. It MAY delete the ticket earlier based on
+            //    local policy.
+            //
+            // RFC 5246 F.1.4 -- TLS 1.2
+            //    If either party suspects that the session may have been
+            //    compromised, or that certificates may have expired or been
+            //    revoked, it should force a full handshake.  An upper limit of
+            //    24 hours is suggested for session ID lifetimes.
+            //
+            // RFC 8446 4.2.11.1 -- TLS 1.3
+            //    The client's view of the age of a ticket is the time since the
+            //    receipt of the NewSessionTicket message.  Clients MUST NOT
+            //    attempt to use tickets which have ages greater than the
+            //    "ticket_lifetime" value which was provided with the ticket.
+            //
+            // RFC 8446 4.6.1 -- TLS 1.3
+            //    Clients MUST NOT cache tickets for longer than 7 days,
+            //    regardless of the ticket_lifetime, and MAY delete tickets
+            //    earlier based on local policy.
+            //
+            // Note: TLS 1.3 tickets with a lifetime longer than 7 days are
+            //       rejected during parsing with an "Illegal Parameter" alert.
+            //       Other suggestions are left to the application via
+            //       Policy::session_ticket_lifetime(). Session lifetimes as
+            //       communicated by the server via the "lifetime_hint" are
+            //       obeyed regardless of the policy setting.
+            const auto session_lifetime_hint = session.lifetime_hint();
+            const bool expired = age > std::min(policy_lifetime, session_lifetime_hint);
+
+            if(expired)
+               { remove(std::get<Session_Handle>(session_and_handle)); }
+
+            return expired;
+            }), sessions_and_handles.end());
+      }
+
+   const auto session_limit = policy.maximum_session_tickets_per_client_hello();
+   while(session_limit > 0 && sessions_and_handles.size() > session_limit)
+      { sessions_and_handles.pop_back(); }
+
+   // RFC 8446 Appendix C.4
+   //    Clients SHOULD NOT reuse a ticket for multiple connections. Reuse of
+   //    a ticket allows passive observers to correlate different connections.
+   //
+   // When reuse of session tickets is not allowed, remove all tickets to be
+   // returned from the implementation's internal storage.
+   if(!policy.reuse_session_tickets())
+      {
+      for(const auto& [session, handle] : sessions_and_handles)
+         {
+         if(!session.version().is_pre_tls_13() || !handle.is_id())
+            {
+            remove(handle);
+            }
+         }
+      }
```

Comment:
Now that this base-implementation of "single-use tickets" utilizes the fact that TLS implementations use the `find()` method atomically we should guard the base implementation with a mutex.

This should be independent of the thread safety requirement of the derived managers, though. Their overrides might be called directly, and hence need to be mutexed as well. However, we could provide a protected interface to reuse the base-classes mutex.

- [ ] Make `Session_Manager` base class thread safe

Reply:
Initially, I wanted to make `Session_Manager` implement the C++11 `Lockable` requirement and let derived classes reuse the base class' mutex simply by: `Botan::lock_guard_type lk(*this)`. Unfortunately, CTAD for aliased templates (`Botan::lock_guard_type`) introduced in C++20 doesn't seem to be implemented by clang (yet?). So one would have explicitly needed to state `Botan::lock_guard_type<Session_Manager>`.

All of that felt like magic and eventually I just exposed `Session_Manager::m_mutex` via a protected accessor for derived classes to use.